Adds an interface for replacing prompt of an individual option in an `OptionList`

nekeal opened the pull request:
**Please review the following checklist.**

- [x] Docstrings on all new or modified functions / classes 
- [ ] Updated documentation
- [x] Updated CHANGELOG.md (where appropriate)

This PR extends `OptionList` class with an interface for replacing prompts in individual options. 

Closes #2603

BenMarkFrost commented on src/textual/widgets/_option_list.py:
```diff
@@ -647,6 +655,65 @@ def remove_option_at_index(self, index: int) -> Self:
+        Raises:
+            IndexError: If there is no option of the given index.
+        """
+        option = self._options[index]
```
There is a get_option_at_index() function already, could we use that here instead for consistency?

```suggestion
        option = self.get_option_at_index(index)
```

rodrigogiraoserrao replied:
That's a good idea, for sure.

rodrigogiraoserrao replied:
Good catch.

BenMarkFrost replied:
This PR is really good, implements the feature required in the issue very well.

I feel like we can do better to avoid rewriting code that already exists here. What do you think @nekeal ?

nekeal replied:
@BenMarkFrost  I pushed the changes. Should I replace it also in the `_remove_option` method?

nekeal replied:
Which part of the code do you mean?

BenMarkFrost replied:
> @BenMarkFrost I pushed the changes. Should I replace it also in the `_remove_option` method?

Yes well spotted, I think that would be a good idea. I can also see it in `_set_option_disabled()`




> Which part of the code do you mean? The one that retrieves the option's index?

This PR re-implements errors for retrieving options by index and ID, when they are already present in those dedicated functions:

`get_option()` (which gets the option by option_id), and `get_option_by_index()`.

------

The implementation in my opinion could be as simple as

```sh
def replace_option_prompt_at_index(index: int .......):
    self.get_option_at_index(index).set_prompt(new_prompt)
    
def replace_option_prompt(id: str .......):
    self.get_option(id).set_prompt(new_prompt)
```

What do you think?

nekeal replied:
Sounds good to me :rocket: 

Should I also change implementations of other methods so they use `get_option()` and `get_option_by_index()`.

BenMarkFrost replied:
Only if you have time / feel like it would be in scope, I think we can happily create another issue for that if it's too much work!

nekeal replied:
I just pushed a [commit](https://github.com/Textualize/textual/pull/2985/commits/1c9f079b5da2c1658e2056d485307fa66f25563b) which shows how the rest of the methods can be implemented. Can you have a look?

rodrigogiraoserrao reviewed (approved):
Looks good to me!
Let's see what CI thinks :P

rodrigogiraoserrao reviewed (changes requested):
Consider addressing https://github.com/Textualize/textual/pull/2985/files#r1271291408.

BenMarkFrost commented on tests/option_list/test_option_prompt_replacement.py:
```diff
@@ -0,0 +1,41 @@
+"""Test replacing options prompt from an option list."""
```
This testing file is excellent 🚀

rodrigogiraoserrao commented on src/textual/widgets/_option_list.py:
```diff
@@ -789,6 +842,22 @@ def get_option(self, option_id: str) -> Option:
             ) from None
 
+    def get_option_index(self, option_id):
+        """
+        Get the index of the option with the given ID.
+
+        Args:
+            option_id: The ID of the option to get the index of.
+        Raises:
+            OptionDoesNotExist: If no option has the given ID.
+        """
```
Very minor, but please tidy the docstring a bit so it looks like the others :)

```suggestion
        """Get the index of the option with the given ID.

        Args:
            option_id: The ID of the option to get the index of.

        Raises:
            OptionDoesNotExist: If no option has the given ID.
        """
```

rodrigogiraoserrao commented on src/textual/widgets/_option_list.py:
```diff
@@ -647,6 +655,51 @@ def remove_option_at_index(self, index: int) -> Self:
+            prompt: The new prompt for the option.
+
+        Returns:
+            The `OptionList` instance.
```
Can we also document what gets raised if the option doesn't exist?

nekeal replied:
Is it necessary if it's already documented in the `get_option_index()` method?

rodrigogiraoserrao replied:
I would say "yes" because users will see the documentation for `replace_option_prompt` and they won't look at how it is implemented to figure out what may get raised.
This will also make sure this method is more aligned with all others, because all other methods say they will raise the error. It's just that we raise it a bit later.

nekeal replied:
Ok, you are right. I added `Raises` section.

rodrigogiraoserrao reviewed (changes requested):
Minor docstring changes, please!

rodrigogiraoserrao commented:
@willmcgugan I'd appreciate your input on this.
Not sure how we feel about the added level of indirection, which can be seen if we compare the `replace_` methods directly with the `remove_` methods, for example.

Related to the discussion in #2986

willmcgugan reviewed (approved):
LGTM! I'd like @davep  to have a quick look, as he wrote the original code...

davep commented:
> I'd like @davep to have a quick look, as he wrote the original code...

Will do. Likely won't be until tomorrow at the earliest, or Monday morning when I get to my desk at the latest. Very keen to look it over.

nekeal commented:
> Will do. Likely won't be until tomorrow at the earliest, or Monday morning when I get to my desk at the latest. Very keen to look it over.

There is not much to look at.  Don't let me leave sprints without a single merged PR ;)

willmcgugan commented:
LGTM. If tests pass, merge away!

rodrigogiraoserrao commented on src/textual/widgets/_option_list.py:
```diff
@@ -789,6 +842,22 @@ def get_option(self, option_id: str) -> Option:
+        """
+        try:
+            return self._option_ids[option_id]
+        except:
```
Why the bare `except` clause?

nekeal replied:
No idea why I missed it. It's limited to `KeyError` now.

rodrigogiraoserrao commented on src/textual/widgets/_option_list.py:
```diff
@@ -647,6 +655,51 @@ def remove_option_at_index(self, index: int) -> Self:
+            prompt: The new prompt for the option.
+
+        Raises:
+            IndexError: If there is no option of the given index.
```
I believe this is incorrect, right? What `self.get_option_at_index` might raise is `OptionDoesNotExist`.

nekeal replied:
Nice catch! Thank you for pointing that out. It's fixed now.

davep commented on src/textual/widgets/_option_list.py:
```diff
@@ -789,6 +842,22 @@ def get_option(self, option_id: str) -> Option:
                 f"There is no option with an ID of '{option_id}'"
             ) from None
 
+    def get_option_index(self, option_id):
```
If we're introducing this as part of the public interface, and in service of `replace_option_prompt`, I think I would reduce the duplication of code by tweaking `remove_option` to be based on it too. I think it makes sense to deduplicate the code in any developer-facing method that's unlikely to be called frequently.

davep replied:
I should mention, same for other similar developer-facing methods. If that's what #2986 was in reference to them I'm in favour of that and would support it.

davep commented on tests/option_list/test_option_prompt_replacement.py:
Nice testing.

One thing I would have done here and I think we should do this is take into account that a prompt can span multiple lines. So I'd write some extra replacement tests where a prompt goes from more to fewer lines, and from fewer to more lines.

I'd also add some snapshot tests too, perhaps:

1. Going from a single-line prompt to a different single-line prompt.
2. Going from many lines to fewer.
3. Going from one or more lines to more lines.

davep reviewed (approved):
This looks great! Hits the right interface, good testing, attention to the docs. Works for me!

I've suggested some small improvements, mainly revolving around making the testing as comprehensive as possible.

rodrigogiraoserrao commented:
Hey @nekeal, I see you replied to my "request changes" but you haven't committed anything. I'm wondering if that's intentional 🤔 
Either way, when you do commit your changes, please let us know if you intend to address @davep's suggestions or not!

Excellent work, so far!

nekeal commented:
> Hey @nekeal, I see you replied to my "request changes" but you haven't committed anything. I'm wondering if that's intentional thinking Either way, when you do commit your changes, please let us know if you intend to address @davep's suggestions or not!
> 
> Excellent work, so far!

No, it wasn't intentional :sweat_smile: Just slow WiFi in a train and I didn't notice that `push` failed. 

Yeah, I will try to implement suggestions from @davep later today but I'm not sure if I'll get it working before the sprint ends. Are you ok with that?

rodrigogiraoserrao commented:
Hey @nekeal, sorry for the time it took me to reply.
Supposedly, I had replied already but apparently I didn't – instead, I edited your last comment and added my reply there...
Which makes no sense 🤣

This is my original reply:

> Yeah, I will try to implement suggestions from @davep later today but I'm not sure if I'll get it working before the sprint ends. Are you ok with that?

Absolutely. The sprint is an event to help people get started with in-person guidance. You are more than welcome to finish the PR “outside” the sprint, as you are absolutely more than welcome to keep contributing.

---

I see you committed something else in the meantime.
Are you happy for me to go over the PR again or do you need more time to finish addressing the feedback?

nekeal commented:
> Hey @nekeal, sorry for the time it took me to reply.
> Supposedly, I had replied already but apparently I didn't – instead, I edited your last comment and added my reply there...
> Which makes no sense 🤣
> 
> This is my original reply:
> 
> > Yeah, I will try to implement suggestions from @davep later today but I'm not sure if I'll get it working before the sprint ends. Are you ok with that?
> 
> Absolutely. The sprint is an event to help people get started with in-person guidance. You are more than welcome to finish the PR “outside” the sprint, as you are absolutely more than welcome to keep contributing.
> 
> ---
> 
> I see you committed something else in the meantime.
> Are you happy for me to go over the PR again or do you need more time to finish addressing the feedback?

No problem at all 😅 
I still need some time to add snapshot tests. I will l let you know when I'm finished.

rodrigogiraoserrao commented:
Ok. When you are ready, feel free to re-request a review by clicking the arrows 🔄 next to my name at the top right.

nekeal commented:
I've added some snapshot tests. I'm not sure if I should've created a new application for testing this particular case but let me know if I should reuse the existing one.

rodrigogiraoserrao reviewed (approved):
Looks good to me!

Sorry I took so long to review but I was away.

rodrigogiraoserrao commented:
Hey @nekeal we're ready to merge this.
The only thing left to do is make sure the changelog is updated correctly.
Would you mind merging `main` into your branch and then making sure your changelog entry ends up at the top?

Thank you _so_ much :D

nekeal commented:
> Hey @nekeal we're ready to merge this. The only thing left to do is make sure the changelog is updated correctly. Would you mind merging `main` into your branch and then making sure your changelog entry ends up at the top?
> 
> Thank you _so_ much :D

No, problem. The branch is rebased with `main` now and hopefully ready to be merged :rocket:

rodrigogiraoserrao commented:
@davep this has now been merged; I believe you wanted to use this elsewhere?

davep commented:
Aye, will merge main into the command palette branch. Thanks all.
